### PR TITLE
Implement traceable torch.tensor when you have SymInt/SymFloat inputs

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109515
* #109513

I just ported the C++ torch.tensor implementation to Python, swapping out the inner bits to successively stack tensors together, so that we can trace through `scalar_tensor`.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
 log_normal_ = _make_inplace(log_normal)
 zero_ = _make_inplace(zero)
 
+
+def _isStorage(obj):
+    # TODO: this is slightly inaccurate; we don't allow subclasses on
+    # UntypedStorage
```
Comment (voznesenskym) [inline torch/_refs/__init__.py:6134]: why not type check directly?

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
...
+def tensor(data, *, dtype=None, device=None, pin_memory=False, requires_grad=False):
+    # TODO (or not): support names kwarg
+    if isinstance(data, torch.Tensor):
+        warnings.warn(
+            "To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() "
+            "or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor)"
+        )
+    type_inference = dtype is None
+    new_tensor = _internal_new_from_data(
+        {"device": "cpu"},  # TODO: use torch.get_default_tensor_type
```
Comment (voznesenskym) [inline torch/_refs/__init__.py:6295]: why device cpu??

Comment (ezyang) [reply torch/_refs/__init__.py:6295]: Because that's the default device, when you don't specify device

File: torch/_refs/__init__.py [hunk 3]
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
 log_normal_ = _make_inplace(log_normal)
 zero_ = _make_inplace(zero)
 
+
+def _isStorage(obj):
```
Comment (voznesenskym) [inline torch/_refs/__init__.py:6132]: hypernit camelcase

File: torch/_refs/__init__.py (hunk 1 above)
Comment (ezyang) [reply torch/_refs/__init__.py:6134]: Hmm, so, originally I was Dynamo'ing through this code and Dynamo didn't support the type() is ... test. But then I realized I didn't need to Dynamo through it I could make_fx trace it instead, probably can fix it now

File: torch/_refs/__init__.py [hunk 4]
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
...
+    type_inference,
+    pin_memory=False,
+):
+    if isinstance(data, torch.Tensor):
+        torch._check(
+            not pin_memory, lambda: "Can't pin tensor constructed from a variable"
+        )
+        var = data
+        if copy_variables:
+            var = var.detach()
```
Comment (voznesenskym) [inline torch/_refs/__init__.py:6243]: sorry, why detach and not clone?

**Review by voznesenskym — APPROVED** [review]
stamp because port but questions

File: torch/_refs/__init__.py [hunk 5]
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
...
+        if length == 0:
+            return torch.get_default_dtype()
+        for i in range(length):
+            cur_item = obj[i]
+            # TODO: test this
+            """
+            if cur_item is obj:
+                raise TypeError("new(): self-referential lists are incompatible")
+            """
+            item_scalarType = _infer_scalar_type(cur_item)  # recurse!
```
Comment (voznesenskym) [inline torch/_refs/__init__.py:6197]: I know its ported but camelcase

File: torch/_refs/__init__.py [hunk 6]
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
...
+            if scalarType is not None:
+                scalarType = torch.promote_types(scalarType, item_scalarType)
```
Comment (voznesenskym) [inline torch/_refs/__init__.py:6199]: more camels

File: torch/_refs/__init__.py [hunk 7]
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
...
+            if scalarType is torch.cdouble:
+                # this won't change (unless we hit undefined, but that will
+                # fail later)
+                return scalarType
+        return scalarType
+    raise RuntimeError(f"Could not infer dtype of {type(obj).__name__}")
+
+
+# Analogous to recursive_store
+def _recursive_build(sizes, dim, scalarType, obj):
```
Comment (voznesenskym) [inline torch/_refs/__init__.py:6211]: assert dim < sizes?

File: torch/_refs/__init__.py [hunk 8]
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
...
+    # TODO (or not): support names kwarg
+    if isinstance(data, torch.Tensor):
+        warnings.warn(
+            "To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() "
+            "or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor)"
+        )
+    type_inference = dtype is None
+    new_tensor = _internal_new_from_data(
+        {"device": "cpu"},  # TODO: use torch.get_default_tensor_type
+        dtype if dtype is not None else torch.get_default_dtype(),
```
Comment (voznesenskym) [inline torch/_refs/__init__.py:6296]: nit: type_inference bool instead of `dtype is not None`

File: torch/_refs/__init__.py [hunk 9]
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
...
+    pin_memory=False,
+):
+    if isinstance(data, torch.Tensor):
+        torch._check(
+            not pin_memory, lambda: "Can't pin tensor constructed from a variable"
+        )
+        var = data
+        if copy_variables:
+            var = var.detach()
+        inferred_scalar_type = var.dtype if type_inference else scalar_type
```
Comment (voznesenskym) [inline torch/_refs/__init__.py:6244]: this is the same line as `inferred_scalar_type = _infer_scalar_type(data) if type_inference else scalar_type` below because that line routes to `.dtype` anyway - combine and move up?

File: torch/_refs/__init__.py (hunk 4 above)
Comment (ezyang) [reply torch/_refs/__init__.py:6243]: you know, I have no idea why the og code does that.  The logic was added in https://github.com/pytorch/pytorch/pull/14097 but we didn't discuss it in review. If I had to guess, it is to deal with the situation if var requires_grad=True, on the inside we might do a .to() call and we want to avoid generating a backwards graph when this happens.

**Comment by ezyang** [comment]
I resolved all the PR comments that would not have caused divergence from the original code
